File: src/run_tests.js
import { exec } from 'child_process';
import { promisify } from 'util';
import * as fs from 'fs';
import * as path from 'path';
import { getRoot } from './path-resolver.js';

const execAsync = promisify(exec);

const TEST_VIDEOS = [
    {
        name: 'test1.mp4',
        description: 'CGI/Animation',
        expectedType: 'cgi',
        required: true,
    },
    {
        name: 'test_drone.mp4',
        description: 'Aerial/Drone Footage',
        expectedType: 'drone',
        required: false,
    },
    {
        name: 'test_anime.mp4',
        description: '2D Animation/Anime',
        expectedType: 'animation',
        required: false,
    },
    {
        name: 'test_vlog.mp4',
        description: 'Handheld Multi-cut Vlog',
        expectedType: 'live-action',
        required: false,
    },
];

const PROJECT_ROOT = getRoot();
const RESULTS_DIR = path.join(PROJECT_ROOT, 'test_results');
const SUMMARY_FILE = path.join(RESULTS_DIR, 'test_summary.json');

function ensureResultsDir() {
    if (!fs.existsSync(RESULTS_DIR)) {
        fs.mkdirSync(RESULTS_DIR, { recursive: true });
    }
}

async function runPipeline(videoPath, options = {}) {
    const timestamp = Date.now();
    const args = [
        'node',
        'src/pipeline.js',
        videoPath,
        '--output-dir', RESULTS_DIR,
    ];

    if (options.dryRun) args.push('--dry-run');
    if (options.model) args.push('--model', options.model);
    if (options.verbose) args.push('--verbose');

    const cmd = args.join(' ');
    console.log(`\n  Running: ${cmd}\n`);

    try {
        const { stdout, stderr } = await execAsync(cmd, {
            cwd: PROJECT_ROOT,
            maxBuffer: 50 * 1024 * 1024,
            timeout: 300000,
        });

        return {
            success: true,
            stdout,
            stderr,
            timestamp,
        };
    } catch (error) {
        return {
            success: false,
            stdout: error.stdout || '',
            stderr: error.stderr || '',
            error: error.message,
            timestamp,
        };
    }
}

function validateOutput(filename) {
    try {
        const content = fs.readFileSync(filename, 'utf-8');
        const data = JSON.parse(content);

        const checks = {
            has_video_metadata: !!data.video_metadata,
            has_blueprint: !!data.blueprint,
            has_prompts: !!data.prompts,
            has_global_aesthetic: !!data.blueprint?.global_aesthetic,
            has_chronological_shots: Array.isArray(data.blueprint?.chronological_shots),
            has_model_outputs: Object.keys(data.prompts || {}).length > 0,
        };

        const passed = Object.values(checks).every(v => v === true);
        const score = (Object.values(checks).filter(v => v).length / Object.keys(checks).length) * 100;

        return { passed, score, checks, data };
    } catch (error) {
        return {
            passed: false,
            score: 0,
            error: error.message,
            checks: {},
        };
    }
}

async function runTests() {
    console.log('═══════════════════════════════════════════');
    console.log('  VideoReverse — Test Suite');
    console.log('═══════════════════════════════════════════\n');

    ensureResultsDir();

    const results = {
        timestamp: new Date().toISOString(),
        total: TEST_VIDEOS.length,
        passed: 0,
        failed: 0,
        tests: [],
    };

    for (const test of TEST_VIDEOS) {
        const videoPath = path.join(PROJECT_ROOT, test.name);
        const exists = fs.existsSync(videoPath);

        console.log(`\n┌─ Test: ${test.name}`);
        console.log(`│  Description: ${test.description}`);
        console.log(`│  Expected Type: ${test.expectedType}`);
        console.log(`│  Required: ${test.required ? 'YES' : 'NO'}`);

        if (!exists) {
            console.log(`│`);
            console.log(`└─ ⏭️  SKIPPED (file not found)`);
            results.tests.push({
                name: test.name,
                status: 'skipped',
                reason: 'file not found',
                required: test.required,
            });
            if (test.required) {
                results.failed++;
            }
            continue;
        }

        const startTime = Date.now();
        const result = await runPipeline(videoPath, { verbose: false });
        const duration = Date.now() - startTime;

        if (result.success) {
            const outputFiles = fs.readdirSync(RESULTS_DIR)
                .filter(f => f.startsWith(test.name.replace('.mp4', '')))
                .map(f => path.join(RESULTS_DIR, f));

            const jsonFile = outputFiles.find(f => f.endsWith('.json'));
            const txtFile = outputFiles.find(f => f.endsWith('.txt'));

            let validation = null;
            if (jsonFile) {
                validation = validateOutput(jsonFile);
            }

            if (validation?.passed) {
                console.log(`│`);
                console.log(`└─ ✅ PASSED (${(duration / 1000).toFixed(1)}s, score: ${validation.score}%)`);
                results.passed++;
                results.tests.push({
                    name: test.name,
                    status: 'passed',
                    duration_ms: duration,
                    validation,
                    output_files: { json: jsonFile, txt: txtFile },
                });
            } else {
                console.log(`│`);
                console.log(`└─ ❌ FAILED (validation score: ${validation?.score || 0}%)`);
                results.failed++;
                results.tests.push({
                    name: test.name,
                    status: 'failed',
                    reason: validation?.error || 'validation failed',
                    duration_ms: duration,
                    validation,
                });
            }
        } else {
            console.log(`│`);
            console.log(`└─ ❌ FAILED (error: ${result.error})`);
            results.failed++;
            results.tests.push({
                name: test.name,
                status: 'failed',
                error: result.error,
                duration_ms: duration,
            });
        }
    }

    console.log('\n═══════════════════════════════════════════');
    console.log('  Test Summary');
    console.log('═══════════════════════════════════════════');
    console.log(`  Total:  ${results.total}`);
    console.log(`  Passed: ${results.passed} ✅`);
    console.log(`  Failed: ${results.failed} ❌`);
    console.log(`  Skipped: ${results.tests.filter(t => t.status === 'skipped').length}`);
    console.log('═══════════════════════════════════════════\n');

    fs.writeFileSync(SUMMARY_FILE, JSON.stringify(results, null, 2));
    console.log(`  Full report: ${SUMMARY_FILE}`);

    return results;
}

if (process.argv[1]?.includes('run_tests')) {
    runTests()
        .then((results) => {
            const exitCode = results.failed > 0 ? 1 : 0;
            process.exit(exitCode);
        })
        .catch((err) => {
            console.error('Test suite error:', err);
            process.exit(1);
        });
}

export { runTests, TEST_VIDEOS };

File: src/run_tests.py
import asyncio
import json
import os
import sys
import time
from datetime import UTC, datetime, timezone
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from src.path_resolver import get_root
from src.pipeline import run_pipeline

TEST_VIDEOS = [
    {
        "name": "test1.mp4",
        "description": "CGI/Animation",
        "expected_type": "cgi",
        "required": True,
    },
    {
        "name": "test_drone.mp4",
        "description": "Aerial/Drone Footage",
        "expected_type": "drone",
        "required": False,
    },
    {
        "name": "test_anime.mp4",
        "description": "2D Animation/Anime",
        "expected_type": "animation",
        "required": False,
    },
    {
        "name": "test_vlog.mp4",
        "description": "Handheld Multi-cut Vlog",
        "expected_type": "live-action",
        "required": False,
    },
]

PROJECT_ROOT = get_root()
RESULTS_DIR = os.path.join(PROJECT_ROOT, "test_results")
SUMMARY_FILE = os.path.join(RESULTS_DIR, "test_summary.json")


def _ensure_results_dir() -> None:
    os.makedirs(RESULTS_DIR, exist_ok=True)


def _validate_output(filename: str) -> dict[str, Any]:
    try:
        with open(filename, encoding="utf-8") as f:
            data = json.load(f)

        checks = {
            "has_video_metadata": bool(data.get("video_metadata")),
            "has_blueprint": bool(data.get("blueprint")),
            "has_prompts": bool(data.get("prompts")),
            "has_global_aesthetic": bool(data.get("blueprint", {}).get("global_aesthetic")),
            "has_chronological_shots": isinstance(data.get("blueprint", {}).get("chronological_shots"), list),
            "has_model_outputs": len(data.get("prompts", {})) > 0,
        }

        passed = all(checks.values())
        score = (sum(1 for v in checks.values() if v) / len(checks)) * 100

        return {"passed": passed, "score": score, "checks": checks, "data": data}
    except Exception as error:
        return {
            "passed": False,
            "score": 0,
            "error": str(error),
            "checks": {},
        }


async def _run_pipeline(video_path: str, options: dict[str, Any] | None = None) -> dict[str, Any]:
    if options is None:
        options = {}
    opts = {
        "video_path": video_path,
        "output_dir": RESULTS_DIR,
        "format": "both",
        "models": options.get("model"),
        "dry_run": options.get("dry_run", False),
        "verbose": options.get("verbose", False),
        "log_level": "debug" if options.get("verbose") else "info",
        "quiet": False,
        "force": False,
        "max_retries": 3,
        "max_duration": None,
        "sample_mode": "full",
        "video_type": None,
        "no_cache": False,
        "wsl_mode": None,
    }

    try:
        result = await run_pipeline(opts)
        return {"success": True, "result": result}
    except Exception as error:
        return {"success": False, "error": str(error)}


async def run_tests() -> dict[str, Any]:
    print("═" * 60, flush=True)
    print("  VideoReverse — Test Suite", flush=True)
    print("═" * 60 + "\n", flush=True)

    _ensure_results_dir()

    results: dict[str, Any] = {
        "timestamp": datetime.now(UTC).isoformat(),
        "total": len(TEST_VIDEOS),
        "passed": 0,
        "failed": 0,
        "tests": [],
    }

    for test in TEST_VIDEOS:
        test_name = str(test["name"])
        video_path = os.path.join(PROJECT_ROOT, test_name)
        exists = os.path.exists(video_path)

        print(f"\n┌─ Test: {test['name']}", flush=True)
        print(f"│  Description: {test['description']}", flush=True)
        print(f"│  Expected Type: {test['expected_type']}", flush=True)
        print(f"│  Required: {'YES' if test['required'] else 'NO'}", flush=True)

        if not exists:
            print("│", flush=True)
            print("└─ ⏭️  SKIPPED (file not found)", flush=True)
            results["tests"].append(
                {
                    "name": test["name"],
                    "status": "skipped",
                    "reason": "file not found",
                    "required": test["required"],
                }
            )
            if test["required"]:
                results["failed"] += 1
            continue

        start_time = time.time() * 1000
        result = await _run_pipeline(video_path, {"verbose": False})
        duration = time.time() * 1000 - start_time

        if result["success"]:
            base_name = test_name.replace(".mp4", "")
            output_files = [os.path.join(RESULTS_DIR, f) for f in os.listdir(RESULTS_DIR) if f.startswith(base_name)]

            json_file = next((f for f in output_files if f.endswith(".json")), None)
            txt_file = next((f for f in output_files if f.endswith(".txt")), None)

            validation = None
            if json_file:
                validation = _validate_output(json_file)

            if validation and validation["passed"]:
                print("│", flush=True)
                print(f"└─ ✅ PASSED ({duration / 1000:.1f}s, score: {validation['score']:.0f}%)", flush=True)
                results["passed"] += 1
                results["tests"].append(
                    {
                        "name": test["name"],
                        "status": "passed",
                        "duration_ms": duration,
                        "validation": validation,
                        "output_files": {"json": json_file, "txt": txt_file},
                    }
                )
            else:
                print("│", flush=True)
                score = validation["score"] if validation else 0
                reason = validation.get("error", "validation failed") if validation else "no output"
                print(f"└─ ❌ FAILED (validation score: {score:.0f}%)", flush=True)
                results["failed"] += 1
                results["tests"].append(
                    {
                        "name": test["name"],
                        "status": "failed",
                        "reason": reason,
                        "duration_ms": duration,
                        "validation": validation,
                    }
                )
        else:
            print("│", flush=True)
            print(f"└─ ❌ FAILED (error: {result['error']})", flush=True)
            results["failed"] += 1
            results["tests"].append(
                {
                    "name": test["name"],
                    "status": "failed",
                    "error": result["error"],
                    "duration_ms": duration,
                }
            )

    skipped = sum(1 for t in results["tests"] if t["status"] == "skipped")

    print("\n" + "═" * 60, flush=True)
    print("  Test Summary", flush=True)
    print("═" * 60, flush=True)
    print(f"  Total:  {results['total']}", flush=True)
    print(f"  Passed: {results['passed']} ✅", flush=True)
    print(f"  Failed: {results['failed']} ❌", flush=True)
    print(f"  Skipped: {skipped}", flush=True)
    print("═" * 60 + "\n", flush=True)

    with open(SUMMARY_FILE, "w", encoding="utf-8") as f:
        json.dump(results, f, indent=2)
    print(f"  Full report: {SUMMARY_FILE}", flush=True)

    return results


def main() -> None:
    results = asyncio.run(run_tests())
    sys.exit(1 if results["failed"] > 0 else 0)


if __name__ == "__main__":
    main()
